fem/libceed/ceed-assemble.cpp

```diff
@@ -0,0 +1,310 @@
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "ceed-assemble.hpp"
+
+#ifdef MFEM_USE_CEED
+#include "ceedsolvers-utility.h"
+
+int CeedHackReallocArray(size_t n, size_t unit, void *p)
+{
+   *(void **)p = realloc(*(void **)p, n*unit);
+   if (n && unit && !*(void **)p)
+      // LCOV_EXCL_START
+      return CeedError(NULL, 1, "realloc failed to allocate %zd members of size "
+                       "%zd\n", n, unit);
+   // LCOV_EXCL_STOP
+
+   return 0;
+}
+
+#define CeedHackRealloc(n, p) CeedHackReallocArray((n), sizeof(**(p)), p)
+
+int CeedHackFree(void *p)
+{
+   free(*(void **)p);
+   *(void **)p = NULL;
+   return 0;
+}
+
+namespace mfem
+{
+
+int CeedSingleOperatorFullAssemble(CeedOperator op, SparseMatrix *out)
+{
+   int ierr;
+   Ceed ceed;
+   ierr = CeedOperatorGetCeed(op, &ceed); CeedChk(ierr);
+
+   // Assemble QFunction
+   CeedQFunction qf;
+   ierr = CeedOperatorGetQFunction(op, &qf); CeedChk(ierr);
+   CeedInt numinputfields, numoutputfields;
+   ierr= CeedQFunctionGetNumArgs(qf, &numinputfields, &numoutputfields);
+   CeedChk(ierr);
+   CeedVector assembledqf;
+   CeedElemRestriction rstr_q;
+   ierr = CeedOperatorLinearAssembleQFunction(
+             op, &assembledqf, &rstr_q, CEED_REQUEST_IMMEDIATE); CeedChk(ierr);
+
+   CeedInt qflength;
+   ierr = CeedVectorGetLength(assembledqf, &qflength); CeedChk(ierr);
+
+   CeedOperatorField *input_fields;
+   CeedOperatorField *output_fields;
+   ierr = CeedOperatorGetFields(op, &input_fields, &output_fields); CeedChk(ierr);
+
+   // Determine active input basis
+   CeedQFunctionField *qffields;
+   ierr = CeedQFunctionGetFields(qf, &qffields, NULL); CeedChk(ierr);
+   CeedInt numemodein = 0, ncomp, dim = 1;
+   CeedEvalMode *emodein = NULL;
+   CeedBasis basisin = NULL;
+   CeedElemRestriction rstrin = NULL;
+   for (CeedInt i=0; i<numinputfields; i++)
+   {
+      CeedVector vec;
+      ierr = CeedOperatorFieldGetVector(input_fields[i], &vec); CeedChk(ierr);
+      if (vec == CEED_VECTOR_ACTIVE)
+      {
+         ierr = CeedOperatorFieldGetBasis(input_fields[i], &basisin);
+         CeedChk(ierr);
+         ierr = CeedBasisGetNumComponents(basisin, &ncomp); CeedChk(ierr);
+         ierr = CeedBasisGetDimension(basisin, &dim); CeedChk(ierr);
+         ierr = CeedOperatorFieldGetElemRestriction(input_fields[i], &rstrin);
+         CeedChk(ierr);
+         CeedEvalMode emode;
+         ierr = CeedQFunctionFieldGetEvalMode(qffields[i], &emode);
+         CeedChk(ierr);
+         switch (emode)
+         {
+            case CEED_EVAL_NONE:
+            case CEED_EVAL_INTERP:
+               ierr = CeedHackRealloc(numemodein + 1, &emodein); CeedChk(ierr);
```

Comment:
You can include `ceed-backend.h` or we can promote the alloc/realloc functions to `ceed.h` to avoid this

Reply:
I think the problem is that some of these functions are marked with hidden visibility.

Reply:
Ah yes. That still might be changeable if it's desirable.